api/shared/condition_types.go

```diff
@@ -0,0 +1,155 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate:=true
+
+package shared
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Conditions for status in web console
+
+// ConditionDetails used for passing condition information into generic functions
+// e.f. GetDataFromSecret
+type ConditionDetails struct {
+	ConditionNotFoundType   ConditionType
+	ConditionNotFoundReason ConditionReason
+	ConditionErrorType      ConditionType
+	ConditionErrordReason   ConditionReason
+}
+
+// ConditionList - A list of conditions
+type ConditionList []Condition
+
+// Condition - A particular overall condition of a certain resource
+type Condition struct {
+	Type               ConditionType          `json:"type"`
+	Status             corev1.ConditionStatus `json:"status"`
+	Reason             ConditionReason        `json:"reason,omitempty"`
+	Message            string                 `json:"message,omitempty"`
+	LastHeartbeatTime  metav1.Time            `json:"lastHearbeatTime,omitempty"`
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// ConditionType - A summarizing name for a given condition
+type ConditionType string
+
+// ConditionReason - Why a particular condition is true, false or unknown
+type ConditionReason string
+
+// NewCondition - Create a new condition object
+func NewCondition(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) Condition {
+	now := metav1.Time{Time: time.Now()}
+	condition := Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastHeartbeatTime:  now,
+		LastTransitionTime: now,
+	}
+	return condition
+}
+
+// Set - Set a particular condition in a given condition list
+func (conditions *ConditionList) Set(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) {
+	condition := conditions.Find(conditionType)
+
+	// If there isn't condition we want to change, add new one
+	if condition == nil {
+		condition := NewCondition(conditionType, status, reason, message)
+		*conditions = append(*conditions, condition)
+		return
+	}
+
+	now := metav1.Time{Time: time.Now()}
+
+	// If there is different status, reason or message update it
+	if condition.Status != status || condition.Reason != reason || condition.Message != message {
+		condition.Status = status
+		condition.Reason = reason
+		condition.Message = message
+		condition.LastTransitionTime = now
+	}
+	condition.LastHeartbeatTime = now
+}
+
+// Find - Check for the existence of a particular condition type in a list of conditions
+func (conditions ConditionList) Find(conditionType ConditionType) *Condition {
+	for i, condition := range conditions {
+		if condition.Type == conditionType {
```

Comment:
If that would also check for ``status == corev1.ConditionTrue``, I think I would get the point of not using a map, but a dict...

Reply:
nvm